qiskit/providers/aer/openpulse/hamiltonian_model.py

```diff
@@ -0,0 +1,344 @@
+                offset = 0
+                while not done:
+                    offset += 1
+                    if not ham_str[kk + offset].isdigit():
+                        done = True
+                    # In case we hit the end of the string
+                    elif (kk + offset + 1) == len(ham_str):
+                        done = True
+                        offset += 1
+                temp_chan = ham_str[kk:kk + offset]
+                if temp_chan not in channels:
+                    channels.append(temp_chan)
+        channels.sort(key=lambda x: (int(x[1:]), x[0]))
+
+        channel_dict = OrderedDict()
+        for idx, val in enumerate(channels):
+            channel_dict[val] = idx
+
+        self._channels = channel_dict
+
+    def _get_diag_hamiltonian(self):
+        """ Get the diagonal elements of the hamiltonian and get the
+        dressed frequencies and eigenstates
+
+        Raises:
+            Exception: Missing index on channel.
+        """
+        # Get the diagonal elements of the hamiltonian with all the
+        # drive terms set to zero
+        for chan in self._channels:
+            exec('%s=0' % chan)
+
+        # might be a better solution to replace the 'var' in the hamiltonian
+        # string with 'op_system.vars[var]'
+        for var in self._vars:
+            exec('%s=%f' % (var, self._vars[var]))
+
+        H_full = np.zeros(np.shape(self._system[0][0].full()), dtype=complex)
+
+        for hpart in self._system:
+            H_full += hpart[0].full() * eval(hpart[1])
+
+        evals, estates = la.eigh(H_full)
+
+        eval_mapping = []
+        for ii in range(len(evals)):
+            eval_mapping.append(np.argmax(np.abs(estates[:, ii])))
+
+        # Remap eigenvalues and eigenstates
+        evals_mapped = evals.copy()
+        estates_mapped = estates.copy()
+
+        for ii, val in enumerate(eval_mapping):
+            evals_mapped[val] = evals[ii]
+            estates_mapped[:, val] = estates[:, ii]
+
+        self._evals = evals_mapped
+        self._estates = estates_mapped
+        self._h_diag = np.ascontiguousarray(np.diag(H_full).real)
+
+
+def _first_excited_state(qubit_idx, dim_qub):
+    """
+    Returns the vector corresponding to all qubits in the 0 state, except for
+    qubit_idx in the 1 state.
+
+    Assumption: the keys in dim_qub consist exactly of the str version of the int
+                in range(len(dim_qub)). They don't need to be in order, but they
+                need to be of this format
+
+    Parameters:
+        qubit_idx (int): the qubit to be in the 1 state
+
+        dim_qub (dict): a dictionary with keys being qubit index as a string, and
+                        value being the dimension of the qubit
+
+    Returns:
+        vector: the state with qubit_idx in state 1, and the rest in state 0
+    """
+    vector = np.array([1.])
+
+    # iterate through qubits, tensoring on the state
+    for idx, dim in enumerate(dim_qub):
```

Comment:
Btw `enumerate` causes an error here, as for some reason it just iterates over the keys, setting both `idx` and `dim` to the current key. Also, I had it iterating over `idx in range(len(dim_qub))` as this way it made sure that the qubit indexes are iterated over in order.

Also, I just realized my old way of doing it is also liable to raise errors, e.g. if a user specifies qubits `[0,2]` for a 3 qubit device. Maybe need to iterate over the list `dim_qub.keys()` after it has been sorted

Reply:
Fixed by writing as `dim_qub.items()` loop